academy/ibc-dev/ibcapp_checkers.md

```diff
@@ -0,0 +1,341 @@
+
+When done, the `x/checkers` module is effectively IBC enabled and you can continue working with Ignite CLI. Scaffold a packet with:
+
+```sh
+$ ignite scaffold packet score playerAddress wonCount:uint DateAdded GameID --module checkers
+```
+
+This creates some files and modifies others like you saw in the [Adding Packet and Acknowledgement data](./ibcapp_packet.md) section. This packet is named differently from the earlier `ibcTopRank` packet. Its fields have different names too, but if the order and the types of the fields match, IBC will have no problem sending these messages across. Of course by default, the changes made by Ignite have unwanted behaviors, e.g. you are able to send an arbitrary `score` for an arbitrary player to the Leaderboard blockchains.
+
+You can tackle those step by step and implement the application logic in the Checkers blockchain. In addition you must adjust the Leaderboard chain and include in it the same Leaderboard structure you saw in the [Migration](/academy/3-my-own-chain/migration.md) section. You need to touch up a couple of files and the Go compiler will alert you if you missed anything. To save you the trouble, you can clone the [sample](https://github.com/b9lab/cosmos-ibc-docker/tree/main/separate), which implements all the changes presented in this section so you can focus on what is important.
+
+## Adjust Checkers
+
+Start by adjusting and implementing how a `score` packet is sent. This logic is implemented in `x/checkers/keeper/message_server_score.go`: 
+
+```go
+...
+func (k msgServer) SendScore(goCtx context.Context, msg *types.MsgSendScore) (*types.MsgSendScoreResponse, error) {
+    ctx := sdk.UnwrapSDKContext(goCtx)
+
+    // TODO: logic before transmitting the packet
+
+    leaderboard, found := k.GetLeaderboard(ctx)
+
+    if !found {
+        errors.New("Leaderboard not found")
+    }
+
+    // Construct the packet
+    var packet types.ScorePacketData
+
+    found_in_leaderboard := false
+    for i := range leaderboard.Winners {
+        if leaderboard.Winners[i].PlayerAddress == msg.Sender {
+            // found the player, send stats
+            packet.PlayerAddress = leaderboard.Winners[i].PlayerAddress
+            packet.WonCount = leaderboard.Winners[i].WonCount
+            packet.DateAdded = leaderboard.Winners[i].DateAdded
+            packet.GameID = msg.ChannelID
+            found_in_leaderboard = true
+            break
+        }
+    }
+
+    if !found_in_leaderboard {
+        errors.New("Player not found in the leaderboard")
+    }
+
+    ...
+}
+```
+
+As you can see, instead of sending an arbitrary score for an arbitrary player, the system first verifies that the player appears on the leaderboard. Use the `ChannelID` as the unique `GameID`. Now you need to modify the `MsgSendScore` in the `proto/tx.proto` and make it to:
+
+```protobuf
+message MsgSendScore {
+    string sender = 1;
+    string port = 2;
+    string channelID = 3;
```

Comment:
The `channelID` here would be the source channel one would specify to send the IBC packet right, so it would not necessarily be unique on the leaderboard chain, correct?

You could only use counterparty `channelID` as a unique game identifier, but even then what happens when a relayer accidentally creates another channel?